examples/fedpca_examples/dim_reduction/client.py

```diff
@@ -0,0 +1,78 @@
+import argparse
+from pathlib import Path
+from typing import Tuple
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flwr.common import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+from torchvision.transforms import transforms
+
+from examples.fedpca_examples.dim_reduction.mnist_model import MnistNet
+from fl4health.clients.basic_client import BasicClient
+from fl4health.preprocessing.pca_preprocessor import PCAPreprocessor
+from fl4health.utils.dataset import MNISTDataset
+from fl4health.utils.metrics import Accuracy
+from fl4health.utils.sampler import DirichletLabelBasedSampler
+
+
+def get_mnist_dataset(data_dir: Path, train: bool) -> MNISTDataset:
+    transform = transforms.Compose(
+        [
+            transforms.ToTensor(),
+            transforms.Normalize((0.5), (0.5)),
+        ]
+    )
+    return MNISTDataset(data_dir, train=train, transform=transform)
+
+
+class MnistFedPCAClient(BasicClient):
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        batch_size = self.narrow_config_type(config, "batch_size", int)
+        pca_path = Path(self.narrow_config_type(config, "pca_path", str))
+        new_dimension = self.narrow_config_type(config, "new_dimension", int)
+        pca_preprocessor = PCAPreprocessor(pca_path)
+        sampler = DirichletLabelBasedSampler(list(range(10)), sample_percentage=0.6, beta=0.75)
+
+        train_dataset = get_mnist_dataset(data_dir=self.data_path, train=True)
+
+        validation_dataset = get_mnist_dataset(data_dir=self.data_path, train=False)
+
+        train_loader = pca_preprocessor.reduce_dimension(
```

Comment:
I would argue that we don't want to support sampler is the reduce_dimension functionality. Unless I'm mistaken, the sampling doesn't have anything to do with dimensionality reduction. So in the interest of "single responsibility," I think we should just have it take a dataset and produce a dataset (with the dimensionality reduced. We can support sampling in `get_mnist_dataset` and create the dataloader from the dimensionality reduced dataset in this function, which wouldn't be a huge refactor I don't think?

Reply:
Similarly, we could support shuffling and batch_size setting when creating the dataloader, rather than having the `reduce_dimension` function being responsible for that.

Reply:
Yes that makes sense, and I also agree that decoupling the dataloader creation from this method is better from a "single responsibility" perspective. I just want to point out that due to the way PyTorch DataLoader class is designed, the dimension reduction transform won't actually be applied to the data until the user iterate through the dataloader in case you had something different in mind 🙂

Reply:
Gotcha. So it's a lazy application. I think we can still modify the transform of the dataset using this functionality in the same way, just with a separation of the functionality.